Simplify test suite invocation

hseg opened the pull request:
In trying to package rnc2rng, I ran into trouble enabling the tests because I didn't recognize the project expected the nonstandard invocation `python testing.py` (instead of using the testing library as a runner as eg `python -m unittest`)

Investigating further, I found that the testing code was building an entire custom framework that was utterly unnecessary -- branching on the presence of `assertMultilineEqual` (the default behaviour for strings since it was added in 3.1), using custom parametrization instead of using subtests (3.4) and using a custom test suite instead of relying on `unittest` discovering the tests through introspection.

This can probably be improved further, I just pushed what was needed for the MWE.

hseg commented:
I've been working against the 2.7.0 release and had dropped my code in out of frustration. I'll first check that there haven't been fixes since

jombr commented:
@hseg I don't think there have been test framework related changes but there have been some fixes related to accepting more of the grammar.

There is also a test related issue #47 that I was hoping to get around to after finishing up after adding some more support for the grammar.

Thank you for the PR I'll take a look later tonight.

hseg commented:
OK, my PR was still useful, no work has been done on `test.py` since 2.7.0.

One regression this PR introduces is that it looks like only two tests are run, but this is misleading -- `test_golden` runs all the golden tests. I don't know how to improve this output without breaking out `pytest` or something else

hseg commented:
In looking up how to make the test suite amenable to being invoked with a simple `python -m pytest`, I ended up also using `importlib.resources` instead of manually playing with paths ourselves.

The end result is that we now can invoke the tests with a simple `python -m unittest` or `python -m pytest`.

hseg commented:
I've tried marking that this closes #47 as well, but with no success.

Closes: #47

hseg commented:
Hi, is there any reason this is still open? What needs to be done to get this merged?

jombr commented:
Hey @hseg that's my bad I got a little busy and let this sit on the back burner. I can commit to reviewing and hopefully merging this weekend.

The one thing I wasn't familiar with was the golden test practice and I wanted to read up on it. I assume it's some common or modern way people are writing tests. If you have a good reference that might help otherwise I can look it up.

hseg commented:
The wiki page on it has a good discussion https://en.wikipedia.org/wiki/Characterization_test

jombr commented:
@hseg it looks good but I think there might be something with pytest and using sub modules I'm not familiar with.

I'm in a venv and I have `pip install .` rnc2rng and also `pip install pytest`

My pip list
```
(.venv) pip list
Package    Version
---------- -------
appdirs    1.4.4
asarPy     1.0.1
iniconfig  2.1.0
packaging  25.0
pip        25.1.1
pluggy     1.6.0
Pygments   2.19.1
pytest     8.4.0
rnc2rng    2.7.0
rply       0.7.8
setuptools 65.5.0
```

When I run `python -m pytest` (and also unittest)

```
========================================== short test summary info ===========================================
FAILED tests/test_golden.py::TestSuite::test_golden - ValueError: '(my_folders)/rnc2rng/tests/golden/include.rng' must be only a file name
```

<summary> Here is the full error </summary>
<details>
(.venv) python -m pytest
============================================ test session starts =============================================
platform linux -- Python 3.11.9, pytest-8.4.0, pluggy-1.6.0
rootdir: /home/deck/git/review/rnc2rng
collected 2 items                                                                                            

tests/test_golden.py .F                                                                                [100%]

================================================== FAILURES ==================================================
___________________________________________ TestSuite.test_golden ____________________________________________

self = <tests.test_golden.TestSuite testMethod=test_golden>

    def test_golden(self):
        for fn in resources.files(golden).iterdir():
            if fn.suffix != 'rnc':
                continue
            with self.subTest(fn=fn):
                self._is_golden(fn)
        # synthesize a test that reads its input from a URL
        with resources.path(golden, 'include.rnc') as path:
>           self._is_golden(path.absolute().as_uri())

tests/test_golden.py:35: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_golden.py:23: in _is_golden
    expected = resources.read_text(golden, ref).rstrip()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.11/importlib/resources/_legacy.py:25: in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.11/importlib/resources/_legacy.py:80: in read_text
    with open_text(package, resource, encoding, errors) as fp:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.11/importlib/resources/_legacy.py:25: in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.11/importlib/resources/_legacy.py:63: in open_text
    return (_common.files(package) / normalize_path(resource)).open(
                                     ^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path = '(my_folders)/rnc2rng/tests/golden/include.rng'

    def normalize_path(path):
        # type: (Any) -> str
        """Normalize a path by ensuring it is a string.
    
        If the resulting string contains path separators, an exception is raised.
        """
        str_path = str(path)
        parent, file_name = os.path.split(str_path)
        if parent:
>           raise ValueError(f'{path!r} must be only a file name')
E           ValueError: '(my_folders)/rnc2rng/tests/golden/include.rng' must be only a file name

/usr/lib/python3.11/importlib/resources/_legacy.py:39: ValueError
============================================== warnings summary ==============================================
tests/test_golden.py::TestSuite::test_from_string
  /home/deck/git/review/rnc2rng/tests/test_golden.py:12: DeprecationWarning: read_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    src = resources.read_text(golden, 'features.rnc')

tests/test_golden.py::TestSuite::test_from_string
tests/test_golden.py::TestSuite::test_from_string
tests/test_golden.py::TestSuite::test_golden
  /usr/lib/python3.11/importlib/resources/_legacy.py:80: DeprecationWarning: open_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    with open_text(package, resource, encoding, errors) as fp:

tests/test_golden.py::TestSuite::test_from_string
  /home/deck/git/review/rnc2rng/tests/test_golden.py:13: DeprecationWarning: read_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    expected = resources.read_text(golden, 'features.rng').rstrip()

tests/test_golden.py::TestSuite::test_golden
  /home/deck/git/review/rnc2rng/tests/test_golden.py:34: DeprecationWarning: path is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    with resources.path(golden, 'include.rnc') as path:

tests/test_golden.py::TestSuite::test_golden
  /home/deck/git/review/rnc2rng/tests/test_golden.py:23: DeprecationWarning: read_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    expected = resources.read_text(golden, ref).rstrip()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================== short test summary info ===========================================
FAILED tests/test_golden.py::TestSuite::test_golden - ValueError: '(my_folders)/rnc2rng/tests/golden/include.rng' must be only a file name
================================== 1 failed, 1 passed, 7 warnings in 0.21s ===================================
</details>

jombr commented:
It seems like if I get the basename it works
`expected = resources.read_text(golden, os.path.basename(ref)).rstrip()`

and if i modify some files I get my failure
```
========================================== short test summary info ===========================================
FAILED tests/test_golden.py::TestSuite::test_golden - AssertionError: '<?xm[1377 chars]s="">foo</name>\n      <text/>\n    </element>[20 chars]mar>' != '<?xm[1...
================================== 1 failed, 1 passed, 7 warnings in 0.21s ===================================
```

Though it is odd that it shows 1 failed and 1 passed. If I modify many files it still says only one failed.

hseg commented:
16 jun 2025 07:41:32 Joseph Bredahl ***@***.***>:

> 
>      [Imagen]*jombr* left a comment (djc/rnc2rng#53)[https://github.com/djc/rnc2rng/pull/53#issuecomment-2975081859]
>     
> 
> It seems like if I get the basename it works
> *expected = resources.read_text(golden, os.path.basename(ref)).rstrip()*
> 
> and if i modify some files I get my failure
> 
> *========================================== short test summary info ===========================================
> FAILED tests/test_golden.py::TestSuite::test_golden - AssertionError: '<?xm[1377 chars]s="">foo</name>\n      <text/>\n    </element>[20 chars]mar>' != '<?xm[1...
> ================================== 1 failed, 1 passed, 7 warnings in 0.21s ===================================
> *
> Though it is odd that it shows 1 failed and 1 passed. If I modify many files it still says only one failed.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/djc/rnc2rng/pull/53#issuecomment-2975081859], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AAJTW5NTYZJMTB2DI5J3LX33DZDHZAVCNFSM6AAAAAB2ERJBIGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNZVGA4DCOBVHE].
> You are receiving this because you were mentioned.
> [Imagen de rastreo][https://github.com/notifications/beacon/AAJTW5PG3PAXJ2A3DHCXWCT3DZDHZA5CNFSM6AAAAAB2ERJBIGWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVRKQSYG.gif]
> 

I'm not sure what's going on, I'll have to investigate this. I'm also surprised subTest isn't continuing test execution. Also, I'm noticing your error report lists DeprecationWarnings I should address before we merge this - after all, part of the motivation here is to modernize invocations. I'll try looking at this tomorrow.

hseg commented:
Could you try running the tests with `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1`?
I'm not sure it'll help in any way, but if any of these packages contain pytest plugins it'll isolate the test from them.

hseg commented:
I can't reproduce your failure, even in a clean chroot with
```
patchutils                 0.4.2-3
python-autocommand         2.2.2-7
python-build               1.2.2-3
python-installer           0.7.0-10
python-jaraco.collections  5.1.0-1
python-jaraco.context      6.0.1-1
python-jaraco.functools    4.1.0-1
python-jaraco.text         4.0.0-2
python-more-itertools      10.7.0-1
python-packaging           25.0-1
python-platformdirs        4.3.6-2
python-pyproject-hooks     1.2.0-3
python-rply                0.7.8-5
python-setuptools          1:80.9.0-1
python-wheel               0.45.1-1
```

EDIT: I just realized that that clean chroot build was testing the new testsuite against 2.7.0, instead of against master, but I doubt that makes a difference?

hseg commented:
Still trying to figure out the subtest thing, though.

hseg commented:
Yikes, testing locally yields false positives after I've shuffled tests/golden/default*.rng and removed a couple of tests. Something's weird here...

hseg commented:
OK, I figured out why it was skipping all the tests - my filter should've been `fn.suffix!=".rnc"` - note the leading period. I've confirmed this change causes all the tests to run, but I still need to check if your other concerns still occur here. I'm out of time for today, hopefully by the weekend I'll be able to finish with this.

hseg commented:
OK, now it's not skipping everything, I see I've written the subtest part correctly.
However, this revealed I hadn't noticed the API requires source locations to be passed as strings, and in particular doesn't support being passed `Path` objects. I'm inclined to fix the library rather than downcasting in the test.
However, `pathlib` was only added in 3.4 and I'm seeing tests in the code for Python 2? I wrote the tests assuming we're following Python support policy, plus perhaps a couple of versions -- how backwards-compatible do you want these changes?

hseg commented:
OK, fixed the tests. The two additional commits enable the parser to be given a `pathlib.Path` as the reference to the path, but now I realize it is utterly unnecessary -- `test_golden::_is_golden` needs to be passed a string anyway (since it does stringly manipulations on inputs that can be paths or strings), so there's no chance of a `Path` object being passed to the parser.
I'm reverting these changes, but you're free to cherry-pick them if they look interesting to you.
I'm particularly happy with my making the isinstance tests implicit by using `@functools.singledispatch` -- they make the code flow of `rnc2rng::parse` clearer IMO. But for the state prior to these commits, this is overkill.

(for future reference, the old commits are https://github.com/djc/rnc2rng/compare/9dbd11a1381cf4770270de4cc169b987992ec419...8419fd024bf4f0b5c940654ab43de5a44cd81b4e -- I just dropped all the changes they made to `rnc2rng/parse.py` when reverting)

jombr commented:
Oh man coming back to this. Sorry its been so long.

> how backwards-compatible do you want these changes?

I think its reasonable expectation to stay in the supported range from https://devguide.python.org/versions/

As of this comment I think 3.10 -> 3.14 is reasonable with 3.15 on the horizon

This shouldn't be bad but it will take some effort fighting a bit of code rot.

jombr commented:
Thank you for your contribution @hseg
